modules/ml/src/main/java/org/apache/ignite/ml/inference/builder/IgniteDistributedInfModelBuilder.java

```diff
@@ -0,0 +1,356 @@
+     * @param <O> Type of model output.
+     * @return Facade represented by {@link InfModel}.
+     */
+    @Override public <I extends Serializable, O extends Serializable> InfModel<I, Future<O>> build(
+        InfModelReader reader, InfModelParser<I, O> parser) {
+        return new DistributedInfModel<>(ignite, UUID.randomUUID().toString(), reader, parser, instances);
+    }
+
+    /**
+     * Apache Ignite service that makes inference reading requests from the request queue and writing responses to the
+     * response queue. This service is assumed to be deployed in {@link #build(InfModelReader, InfModelParser)} method
+     * and cancelled in {@link InfModel#close()} method of the inference model.
+     *
+     * @param <I> Type of model input.
+     * @param <O> Type of model output.
+     */
+    private static class IgniteDistributedInfModelService<I extends Serializable, O extends Serializable>
+        implements Service {
+        /** */
+        private static final long serialVersionUID = -3596084917874395597L;
+
+        /** Inference model reader. */
+        private final InfModelReader reader;
+
+        /** Inference model parser. */
+        private final InfModelParser<I, O> parser;
+
+        /** Suffix that with correspondent templates formats service and queue names. */
+        private final String suffix;
+
+        /** Request queue, is created in {@link #init(ServiceContext)} method. */
+        private transient IgniteQueue<I> reqQueue;
+
+        /** Response queue, is created in {@link #init(ServiceContext)} method. */
+        private transient IgniteQueue<O> resQueue;
+
+        /** Inference model, is created in {@link #init(ServiceContext)} method. */
+        private transient InfModel<I, O> mdl;
+
+        /**
+         * Constructs a new instance of Ignite distributed inference model service.
+         *
+         * @param reader Inference model reader.
+         * @param parser Inference model parser.
+         * @param suffix Suffix that with correspondent templates formats service and queue names.
+         */
+        IgniteDistributedInfModelService(InfModelReader reader, InfModelParser<I, O> parser, String suffix) {
+            this.reader = reader;
+            this.parser = parser;
+            this.suffix = suffix;
+        }
+
+        /** {@inheritDoc} */
+        @Override public void init(ServiceContext ctx) {
+            Ignite ignite = Ignition.localIgnite();
+
+            reqQueue = ignite.queue(String.format(INFERENCE_REQ_QUEUE_NAME_PATTERN, suffix), QUEUE_CAPACITY, queueCfg);
+            resQueue = ignite.queue(String.format(INFERENCE_RES_QUEUE_NAME_PATTERN, suffix), QUEUE_CAPACITY, queueCfg);
+
+            mdl = parser.parse(reader.read());
+        }
+
+        /** {@inheritDoc} */
+        @Override public void execute(ServiceContext ctx) {
+            while (!ctx.isCancelled()) {
+                I req;
+                try {
+                    req = reqQueue.take();
```

Comment:
Premature optimization:
If "mdl.predict(req);" will be cheap call then we can be slow downed due to parks/unparks on "reqQueue.take()"

Maybe we should create some benchmark for this code

Reply:
I would prefer do not make any optimizations before fundamental queue problems are not solved (see [IGNITE-10250](https://issues.apache.org/jira/browse/IGNITE-10250)).